src/record.rs

```diff
@@ -106,15 +111,46 @@ impl Record {
         &self.header
     }
 
-    /// # Description
-    /// Serialize record to bytes
-    pub fn to_raw(&self) -> bincode::Result<Vec<u8>> {
-        let mut buf = self.header.to_raw()?;
-        trace!("raw header: len: {}", buf.len());
-        let raw_meta = self.meta.to_raw()?;
-        buf.extend(&raw_meta);
+    pub(crate) async fn write_to_file(&self, file: &File) -> Result<u64> {
+        if self.data.len() <= MAX_SINGLE_PASS_DATA_SIZE {
+            self.write_single_pass(file).await
+        } else {
+            self.write_double_pass(file).await
+        }
+    }
+
+    fn create_header_meta_buffer(&self) -> Result<Vec<u8>> {
+        let size = self.header.serialized_size() + self.meta.serialized_size();
+        let mut c = Cursor::new(Vec::with_capacity(size as usize));
+        self.header.to_raw_into(&mut c)?;
+        self.meta.to_raw_into(&mut c)?;
+        Ok(c.into_inner())
+    }
+
+    async fn write_single_pass(&self, file: &File) -> Result<u64> {
+        let mut buf = self.create_header_meta_buffer()?;
         buf.extend(&self.data);
-        Ok(buf)
+        Self::process_file_result(file.write_append(&buf).await)
+    }
+
+    async fn write_double_pass(&self, file: &File) -> Result<u64> {
+        let buf = self.create_header_meta_buffer()?;
+        Self::process_file_result(
+            file.write_append(&buf)
```

Comment:
`file.write_append` call gives overhead due to usage of `spawn_blocking`. It is better to add an additional function that takes 2 buffers and write them. This will additionally allow to use `write_all_vectored` (https://doc.rust-lang.org/std/io/trait.Write.html#method.write_all_vectored) in future